File: backend/app/routes/insights.py
from flask import Blueprint, current_app, jsonify, request

from ..services.graph_analysis_service import build_graph_scope_key
from ..services.job_repository import job_repository
from ..services.task_runner import enqueue_graph_analysis

insights_bp = Blueprint("insights", __name__)


@insights_bp.get("/topics")
def topics_graph():
    category = request.args.get("category", "", type=str)
    book_id = request.args.get("book_id", type=int)
    time_scope = request.args.get("time_scope", "all", type=str)
    mode = request.args.get("mode", "category", type=str)
    scope_key = build_graph_scope_key(
        category=category,
        book_id=book_id,
        time_scope=time_scope,
        mode=mode,
    )
    cached = job_repository.get_graph_analysis(scope_key)
    if cached:
        return jsonify(cached)

    job = enqueue_graph_analysis(
        current_app._get_current_object(),
        {
            "category": category,
            "book_id": book_id,
            "time_scope": time_scope,
            "mode": mode,
        },
    )
    return (
        jsonify(
            {
                "overview": None,
                "filters": None,
                "clusters": [],
                "graph": {"nodes": [], "links": []},
                "status": job["status"],
                "job_id": job["id"],
                "message": job["message"],
            }
        ),
        202,
    )

File: frontend/src/types/insights.ts
export interface TopicGraphOverview {
  topic_count: number
  cluster_count: number
  edge_count: number
  book_count: number
}

export interface TopicGraphFilterOption {
  label: string
  value: string
}

export interface TopicGraphBookOption {
  id: number
  title: string
  category: string
}

export interface TopicGraphBook {
  id: number
  title: string
  cover: string
  notes: number
}

export interface TopicGraphSample {
  note_id: number
  book_id: number
  book_title: string
  excerpt: string
}

export interface TopicClusterAction {
  label: string
  description: string
  path: string
  type: 'qa' | 'notes' | 'review'
}

export interface TopicCluster {
  id: number
  name: string
  topics: string[]
  note_count: number
  book_count: number
  sample_books: TopicGraphBook[]
  sample_excerpts: TopicGraphSample[]
  actions?: TopicClusterAction[]
}

export interface TopicGraphNode {
  id: string
  name: string
  value: number
  note_count: number
  book_count: number
  cluster_id: number
}

export interface TopicGraphLink {
  source: string
  target: string
  value: number
  co_occurrence: number
  shared_books: number
}

export interface TopicGraphPayload {
  overview: TopicGraphOverview | null
  filters: {
    selected: {
      category: string
      book_id: number | null
      time_scope: string
      mode: 'category' | 'topic'
    }
    categories: string[]
    books: TopicGraphBookOption[]
    time_scopes: TopicGraphFilterOption[]
    modes: TopicGraphFilterOption[]
  } | null
  clusters: TopicCluster[]
  graph: {
    nodes: TopicGraphNode[]
    links: TopicGraphLink[]
  }
  status?: 'queued' | 'processing' | 'success'
  job_id?: string
  message?: string
}
